### ISSUE TITLE
Empty schemes array is being rendered, even though protocol is defined in @ApiOperation

### ISSUE THREAD

**Issue by who** [springfox/springfox#701]
I was under the impression that one could set a value for `protocols` in `@ApiOperation` to populate a value for the `schemes` key in the Swagger 2.0 output, but this appears to be not working.

This annotation ...

```
@RestController
@RequestMapping("/bookstore")
@Api(
    value = "this is the api value string in class level annotation",
    description = "API for book store")
public class BookStoreController {
...
    @RequestMapping(
        value = "/books", 
        method = RequestMethod.POST, 
        consumes = MediaType.APPLICATION_JSON_VALUE,       
        produces = MediaType.APPLICATION_JSON_VALUE)
    @ApiOperation(
        value = "Create a book",
        notes = "Creates a book so that it can be managed by this api.",
        response = Book.class,
        protocols = "http"
    )
    public Book createBook(
        @ApiParam(value = "Book object", required = true) @RequestBody @Valid Book book) {
        return bookService.createBook(book);
    }
```

... renders this schema:

```
{
  ...
  "paths": {
    "/books": {
      "post": {
        "tags": [
          "this-is-the-api-value-string-in-class-level-annotation"
        ],
        "summary": "Creates a book",
        "description": "Creates a book so that it can be managed by this api.",
        "schemes": [],
        "consumes": [
          "application/json"
        ],
        "produces": [
          "application/json"
        ],
        "parameters": [
          {
            "in": "body",
            "name": "body",
            "description": "Book object",
            "required": true,
            "schema": {
              "$ref": "#/definitions/Book"
            }
          }
        ],
        "responses": {
          ...
        }
      }
    }
  }
  ...
}
```

Expected: **`... "schemes": ["http"] ... `**

**Comment by dilipkrish** [comment]
true dat! thanks for catching the miss

**Comment by dilipkrish** [comment]
However, its unlikely that a resource operation would have something thats different that the globally specified protocols. I've marked this as a bug to fix in the interest of completeness

**Comment by who** [comment]
Also - when setting a `protocols` value for the class-level `@Api` annotation, the value does not cascade into the methods of that class.

**Comment by who** [comment]
I should also mention that setting the protocols in your Docket does indeed render a value for `schemes` at the ApiListing level.

This docket ... 
```
    ...
    public Docket swaggerSpringMvcPlugin() {        
        return new Docket(DocumentationType.SWAGGER_2)
            .protocols(Sets.newHashSet("http"))
            ....
```

... renders this spec:

```
{
    "swagger": "2.0",
    "info": {
        ...
    },
    "host": "localhost:8080",
    "basePath": "/bookstore",
    "tags": [
        ...
    ],
    "schemes": [
        "http"
    ]
    ...
}
```

**Comment by dilipkrish** [comment]
Like I said earlier, all this is really very theoretical. You're not going to find operations available @ different protocols than the controller, different than the ones supported by the service itself. More importantly these aren't in place because swagger 1.2 never supported it. Since most of the 2.0 work is about preserving the 1.2 contracts, we haven't had a chance to revisit some of these misses. Thanks for catching this! :metal:  

We're kind of getting ready to release 2.0 so we may or may not pull this in. /c @adrianbk

**Comment by who** [comment]
@dilipkrish Yes, I agree that a diversity of protocols at varying levels are uncommon in an API.

However, when `schemes` is defined for a path, even when defined as `[]`, some consumers of the Swagger spec will interpret this as an over-ride of the top-level `schemes` value. This is happening currently with swagger-editor, and is actually how I discovered this issue, by copying springfox output and pasting it into the JSON importer of swagger-editor. Because an empty `schemes` array appears to over-ride the default top-level `schemes` value in swagger-editor, the editor tries to hit `undefined://api.example.com/mybasepath`.

Swagger UI, on the other hand, appears to make the interpretation that an empty schemes array is not an override.

**Comment by who** [comment]
From the "schemes" row of https://github.com/swagger-api/swagger-spec/blob/master/versions/2.0.md#operation-object
> The value overrides the Swagger Object schemes definition.

So, if you have `"schemes":["https"]` defined at the top level and `"schemes":[]` defined at the operation level, if I am interpreting the Swagger 2.0 spec correctly, `[]` will override `["https"]`.

**Comment by dilipkrish** [comment]
You're right it will and that may be swagger-ui doing a best effort at working with the service description.

**Comment by who** [comment]
Fix proposal: 

Perhaps we should not render the `schemes` key at all for a given api operation unless the `@ApiOperation` explicitly defines a value for `protocols`.

**Comment by dilipkrish** [comment]
@who you want help us out with a PR? We're stretched pretty thin

**Comment by who** [comment]
@dilipkrish Will do.  Glad to help.

**Comment by dilipkrish** [comment]
Thanks! Just thinking through it I believe you'll need to create an ```OperationSchemesReader``` in ```springfox-swagger-common``` that reads an ```@ApiOperation#protocols``` attribute. Not sure what it would take to read @Api annotation. That might be harder!

**Comment by adrianbk** [comment]
@dilipkrish @who It's a bug for sure and we should try fix it before releasing. @who I agree with your fix proposal, generally we take the same approach with all annotation customisations: The swagger annotation values, as supplied by the user, overrules what's been inferred by Springfox.

@who do you have some bandwidth to work on this?

**Comment by dilipkrish** [comment]
@who I was just looking at this and I've been unable to reproduce this bug. Looks like its because the object mapper thats being used isnt the same thats been configured. Are you customizing your ```ObjectMapper``` by any chance?

I added the following code to the [feature demonstration service](https://github.com/springfox/springfox/blob/master/springfox-spring-web/src/test/java/springfox/documentation/spring/web/dummy/controllers/FeatureDemonstrationService.java#L113-L117)
```java
  @RequestMapping
  @ApiOperation(value = "all", protocols = "http" /*<-- this */)
  public HttpEntity<String> all() {
    throw new UnsupportedOperationException();
  }
```

I was expecting to see the schemes [empty list show up here](https://github.com/springfox/springfox/blob/master/swagger-contract-tests/src/test/resources/contract/swagger2/declaration-feature-demonstration-service.json) and it sure doesn't. 

Do you have a repo that demonstrates this issue? or could you perhaps modify the tests on this repo to break the test. AFAIK the operation has never had support for schemes so its odd that you'd see an empty list.

**Comment by dilipkrish** [comment]
I vote that this is moved to 2.1 if we arent able to re-create it.

**Comment by adrianbk** [comment]
:+1: for 2.1 if it's not reproducible.

**Comment by who** [comment]
@dilipkrish, you are correct! The issue seems to lie in one of my ObjectMapper customizations:

```
objectMapper.disable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES);
```

When I remove this, I no longer see the empty `"schemes":[]` being rendered in my Swagger 2.0 spec.

**Comment by dilipkrish** [comment]
yaaay!! close?

**Comment by who** [comment]
Yay indeed! Now I just have to figure out to make the ObjectMapper customization work in harmony with Springfox.  Closing.

**Comment by who** [comment]
@dilipkrish See edit of https://github.com/springfox/springfox/issues/701#issuecomment-95376903

Turns out my Hibernate4Module was causing all empty arrays to be rendered in the Swagger 2.0 spec, so even the `consumes`, `produces`, and `protocols` arrays were all rendering as `[]`.  

This is really unfortunate, as my API really needs to display nulls and empty arrays.

**Comment by adrianbk** [comment]
@who, @dilipkrish and I had just being discussing this. I think we will need to isolate the object mappers. There should not be any side effects caused by user customisations to the applications primary ObjectMapper or ObjectMappers used by other libraries,

**Comment by who** [comment]
@adrianbk :+1: That would be a very welcome change.  Springfox ObjectMappers should work inside their own isolated vaccum.

**Comment by dilipkrish** [comment]
@who would you mind forking this repo and replicating this issue using the swagger-contract-tests module?  To me isolating the object mappers feels like a considerable change and veering off the beaten path. So I'd like to understand what the problem is before we make changes to how things work.

**Comment by who** [comment]
@dilipkrish I'll have a go, sure.

**Comment by dilipkrish** [comment]
@who I've add a few tests and I've not been able to break the swagger specific mixin serialization rules. These are ofcourse very simplistic tests that are the bare minimum needed to demonstrate the issue. Feel free to add your own tests as well, or whatever customizations you have and see if it breaks the serialization.

**Comment by adrianbk** [comment]
@dilipkrish @who I've created a failing test which reproduces the issue  https://github.com/springfox/springfox/commit/9bdbf62b7f58154964d29c9083028f7942c36f74

I'm of the opinion we should avoid using ObjectMappers which have been user or library configured to  serialize Swagger JSON.

I'm going to have a go at fixing this - it will involve rewriting the controllers that serve out the swagger JSON to use an ObjectMapper instance that is dedicated to springfox and is not registered with Spring's `MappingJackson2HttpMessageConverter`

**Comment by adrianbk** [comment]
@who can you verify with the snapshot produced by this build:https://circleci.com/gh/springfox/springfox/1006

**Comment by who** [comment]
@dilipkrish @adrianbk Brilliant!

I just pulled the latest code from https://github.com/springfox/springfox/commit/28db032f866a4a65f58479c172d7e22df47913ff , built and installed to my local maven, and confirmed I was indeed able to have ObjectMapper customizations and settings in place while no longer seeing empty values rendered by Springfox.  Nice work!

This also fixes some of the other issues I was having with Springfox - namely I had to keep re-defining my produces and consumes annotation for every RequestMapping, because the ObjectMapper collision was causing Springfox to render empty arrays for them too.

If it proves to be stable, which appears to be the case so far, your change here fixed a whole suite of issues and prevented many future issues from being reported.  Kudos - you guys do excellent work.
